src/daemon/utils_hashtable.c

```diff
@@ -0,0 +1,474 @@
+#include <stdint.h>
+#include <assert.h>
+#include <errno.h>
+
+
+/*
+ * Implementation notes:
+ *
+ * The size of the table (i.e. the number of buckets) is always a
+ * power of 2.
+ *
+ * We use double hashing where both h1 and h2 are taken from the
+ * user-supplied hash value. The low order bits go into h1, the next
+ * higher ones into h2. The probe sequence is then
+ * h1 + i * h2 for i = 0, 1, 2, ...
+ * h2 needs to be relatively prime to the table size to ensure all
+ * buckets get probed eventually. This is ensured by setting h2's
+ * lowest bit to 1, making it an odd number while the size is a power
+ * of 2.
+ *
+ * To ensure we have some entropy in the higher order bits (so we get
+ * sensible values for h2), the user-supplied hash value is multiplied
+ * by HASH_MULT (a relatively large prime).
+ *
+ * When an entry is deleted, we put a "tombstone" marked by a status of
+ * BUCKET_TOMB in its location. This is to ensure the probe sequence
+ * for other entries does not get interrupted.
+ *
+ * Tombstones must be removed eventually if they grow too many; this
+ * is realized with a rehash: A new table is allocated and all entries
+ * from the old table are reinserted into the new one.
+ *
+ * The same rehash procedure is also used to resize the table. It both
+ * grows and shrinks dynamically according to the number of elements
+ * it contains. For optimum performance, we maintain a load factor
+ * between 1/8 and 1/2 the table size; the number of tombstones is kept
+ * below 1/4 the size.
+ */
+
+
+#define MIN_ALIGN 16
+
+
+enum bucket_status_e {
+    BUCKET_EMPTY    = 0,
+    BUCKET_USED     = 1,
+    BUCKET_TOMB     = 2,
+};
+
+typedef enum bucket_status_e bucket_status_t;
+
+
+struct bucket_s {
+    hash_t hash;
+    bucket_status_t status;
+};
+
+typedef struct bucket_s bucket_t;
+
+
+static inline uintptr_t do_align (uintptr_t base, uintptr_t alignment);
+static inline bucket_t *get_bucket (hashtable_t *table, hash_t index);
+static inline bucket_t *get_bucket2 (char *arena, hash_t stride, hash_t index);
+static int alloc_arena (hashtable_t *table, unsigned size_exp);
+int hashtable_init (hashtable_t *table, unsigned data_size, unsigned alignment, unsigned minsize_exp);
+static inline hash_t mod_size (const hashtable_t *table, hash_t x);
+static inline hash_t get_h1 (const hashtable_t *table, hash_t hash);
+static inline hash_t get_h2 (const hashtable_t *table, hash_t hash);
+static inline void *bucket_to_user (const hashtable_t *table, bucket_t *bucket);
+static inline bucket_t *user_to_bucket (const hashtable_t *table, void *user);
+static int rehash (hashtable_t *table, unsigned size_exp);
+static int check_grow (hashtable_t *table);
+static int check_shrink (hashtable_t *table, bool bulk);
+static int hashtable_lookup_internal (hashtable_t *table, hash_t hash,
+        bool (*match) (void *, void *), void *match_arg, void **data);
+
+
+static inline uintptr_t
+do_align (uintptr_t base, uintptr_t alignment)
```

Comment:
Please don't hand-roll alignment code. Please use the C11 function(s) instead: `aligned_alloc(size_t alignment, size_t size)`